File: src/hero-death-journal.ts
import type {ParsedMatch,HeroCounterWindow,HeroDeathJournalEntry} from './replay.js';
import {heroName} from './player-profile.js';
import {verifiedReplayScoreboard} from './replay-scoreboard.js';
import {assertHeroDeathJournal} from './hero-death-contract.js';

export interface HeroDeathView {
 id:string;seconds:number;status:HeroDeathJournalEntry['status'];victim:number|null;killer:number|null;
 attribution:HeroDeathJournalEntry['attribution'];x:number|null;y:number|null;coordinatesSource:string|null;legacyDeathId:string|null;
 sourceControlled:boolean;sourceIllusion:boolean;deathCounter:HeroCounterWindow|null;killCounter:HeroCounterWindow|null;
}
export interface HeroDeathModel {
 players:{hero:string;heroLabel:string;steamId:string;team:'radiant'|'dire'}[];
 entries:HeroDeathView[];cells:{killer:number;victim:number;count:number|null;eventIds:string[]}[];
 coverage:{complete:boolean;verifiedDeaths:number;verifiedKills:number;reincarnations:number;unresolved:number;nonHeroKills:number;invalid:number};
}
const finite=(n:unknown):n is number=>typeof n==='number'&&Number.isFinite(n);
const integer=(n:unknown):n is number=>Number.isSafeInteger(n)&&Number(n)>=0;
const norm=(s:string)=>s.replace(/^npc_dota_hero_/,'').replace(/_/g,'');
const counter=(c:HeroCounterWindow|null|undefined,seconds:number):c is HeroCounterWindow=>!!c&&c.source==='CDOTA_PlayerResource'&&finite(c.from)&&finite(c.to)&&c.to>=c.from&&seconds>=c.from-.01&&seconds<=c.to+.01&&[c.before,c.after,c.delta].every(integer)&&c.after-c.before===c.delta;
const key=(index:number,c:HeroCounterWindow)=>[index,c.from,c.to,c.before,c.after,c.delta].join(':');

/** Derive rivalries from independently observed counter increments, never from final KDA subtraction. */
export function buildHeroDeathJournal(match:ParsedMatch):HeroDeathModel|null {
 const raw=match.hero_death_journal;
 try{assertHeroDeathJournal(raw);}catch{return null;}
 const c=raw.coverage;
 if(!Array.isArray(match.players))return null;
 const duration=finite(match.apm_duration_seconds)&&match.apm_duration_seconds>0?match.apm_duration_seconds:match.duration_min*60;
 if(!finite(duration)||duration<0)return null;
 const legacyDuration=Math.round(Math.max(0,match.duration_min)*60);
 const players=match.players.map(p=>({hero:p.hero,heroLabel:heroName(p.hero),steamId:p.steam_id,team:p.team}));
 function identity(hero:string|null,steam:string|null,team:string|null,slot:number|null):number|null {
  if(typeof hero!=='string'||(team!=='radiant'&&team!=='dire'))return null;
  const found=match.players.flatMap((p,i)=>{const s=verifiedReplayScoreboard(p);return norm(p.hero)===norm(hero)&&p.team===team&&(!steam||p.steam_id===steam)&&(slot===null||!s||s.resource_slot===slot)?[i]:[];});
  return found.length===1?found[0]:null;
 }
 const seen=new Set<string>(),duplicates=new Set<string>();for(const e of raw.entries){if(e&&typeof e.id==='string'){if(seen.has(e.id))duplicates.add(e.id);seen.add(e.id);}}
 let invalid=0;
 const rows=raw.entries.flatMap(e=>{
  if(!e||typeof e.id!=='string'||duplicates.has(e.id)||!finite(e.seconds)||e.seconds< -3600||e.seconds>duration+.25){invalid++;return [];}
  const candidateVictim=identity(e.victim_hero,e.victim_steam_id,e.victim_team,e.victim_resource_slot),candidateKiller=identity(e.killer_hero,e.killer_steam_id,e.killer_team,e.killer_resource_slot);
  const victim=candidateVictim!==null&&e.raw_target.startsWith('npc_dota_hero_')&&norm(e.raw_target)===norm(players[candidateVictim].hero)?candidateVictim:null;
  const killer=candidateKiller!==null&&e.raw_source?.startsWith('npc_dota_hero_')&&norm(e.raw_source)===norm(players[candidateKiller].hero)&&e.attacker_team===players[candidateKiller].team?candidateKiller:null;
  const d=counter(e.death_counter,e.seconds)?e.death_counter:null,k=counter(e.kill_counter,e.seconds)?e.kill_counter:null;
  let status:HeroDeathView['status']='unverified';
  if(e.status==='contradiction')status='contradiction';
  else if(e.status==='verified_death'&&e.will_reincarnate!==true&&victim!==null&&d&&d.delta>0)status='verified_death';
  else if(e.status==='verified_reincarnation'&&victim!==null&&e.will_reincarnate===true&&d?.delta===0)status='verified_reincarnation';
  const eligibleKiller=status==='verified_death'&&killer!==null&&victim!==null&&players[killer].team!==players[victim].team&&e.attribution==='verified_enemy_kill'&&k&&k.delta>0?killer:null;
  return [{raw:e,view:{id:e.id,seconds:e.seconds,status,victim,killer:eligibleKiller,attribution:eligibleKiller!==null?'verified_enemy_kill':e.attribution==='verified_enemy_kill'?'unknown':e.attribution,x:null,y:null,coordinatesSource:null,legacyDeathId:null,sourceControlled:e.source_controlled===true,sourceIllusion:e.source_illusion===true,deathCounter:d,killCounter:eligibleKiller!==null?k:null} as HeroDeathView}];
 });
 // Every event claiming a shared delta must account for that whole delta exactly once.
 const deaths=new Map<string,typeof rows>(),kills=new Map<string,typeof rows>();
 for(const row of rows){const v=row.view;if(v.status==='verified_death'&&v.victim!==null&&v.deathCounter){const k=key(v.victim,v.deathCounter);deaths.set(k,[...(deaths.get(k)??[]),row]);}if(v.killer!==null&&v.killCounter){const k=key(v.killer,v.killCounter);kills.set(k,[...(kills.get(k)??[]),row]);}}
 // Distinct groups cannot each claim the same counter increment with slightly
 // different observation windows. Keep unrelated, fully proved groups usable.
 function incompleteGroups(groups:Map<string,typeof rows>,field:'deathCounter'|'killCounter'){
  const bad=new Set<typeof rows>(),all=[...groups.values()];
  for(const group of all){const window=group[0].view[field]!;if(group.length!==window.delta)bad.add(group);}
  for(let i=0;i<all.length;i++)for(let j=i+1;j<all.length;j++){
   const a=all[i][0].view,b=all[j][0].view,owner=field==='deathCounter'?'victim':'killer';
   const x=a[field]!,y=b[field]!;
   if(a[owner]===b[owner]&&Math.max(x.before,y.before)<Math.min(x.after,y.after)){bad.add(all[i]);bad.add(all[j]);}
  }
  return bad;
 }
 const badDeaths=incompleteGroups(deaths,'deathCounter'),badKills=incompleteGroups(kills,'killCounter');
 for(const group of badDeaths)for(const {view:v} of group){v.status='contradiction';v.killer=null;v.killCounter=null;v.attribution='unknown';}
 for(const group of kills.values())if(badKills.has(group)||group.some(r=>r.view.status!=='verified_death'))for(const {view:v} of group){v.killer=null;v.killCounter=null;v.attribution='unknown';}
 // Reuse an existing death point/recap only when victim and precise combat time match uniquely.
 const pointCandidates=rows.map(({view:v})=>{
  if(v.victim===null)return [];
  const details=match.players[v.victim].combat_details;
  if(!details||!['combat-log-v1','combat-log-v2','combat-log-v3'].includes(details.version)||!Array.isArray(details.deaths))return [];
  // MatchInsights assigns recap indices after this exact version/time filter.
  const filtered=details.deaths.filter(d=>d&&finite(d.min)&&d.min>=0&&d.min*60<=legacyDuration);
  return filtered.flatMap((d,i)=>Math.abs(d.min*60-v.seconds)<=.031+1e-9?[{death:d,index:i,player:v.victim!}]:[]);
 });
 const pointUses=new Map<string,number>();for(const group of pointCandidates)for(const p of group){const k=p.player+':'+p.index;pointUses.set(k,(pointUses.get(k)??0)+1);}
 rows.forEach(({view:v},i)=>{const group=pointCandidates[i];if(group.length!==1)return;const p=group[0];if(pointUses.get(p.player+':'+p.index)!==1)return;
  if(finite(p.death.x)&&finite(p.death.y)){v.x=p.death.x;v.y=p.death.y;v.coordinatesSource=p.death.coordinates_source??null;}
  if(v.status==='verified_death'&&v.seconds>=0)v.legacyDeathId=match.players[p.player].steam_id+':'+p.index;
 });
 const entries=rows.map(r=>r.view),verifiedDeaths=entries.filter(e=>e.status==='verified_death').length,verifiedKills=entries.filter(e=>e.killer!==null).length,reincarnations=entries.filter(e=>e.status==='verified_reincarnation').length;
 const unresolved=entries.filter(e=>e.status==='unverified'||e.status==='contradiction'||e.status==='verified_death'&&e.killer===null&&e.attribution==='unknown').length;
 const rosterHeroes=new Set(players.map(p=>norm(p.hero))),rosterAccounts=new Set(players.map(p=>p.steamId));
 const auditHeroes=new Set(c.final_audit.map(a=>norm(a.hero)));
 const audit=c.final_audit.length===players.length&&rosterHeroes.size===players.length&&rosterAccounts.size===players.length&&auditHeroes.size===players.length&&c.final_audit.every(a=>{
  const index=players.findIndex(p=>norm(p.hero)===norm(a.hero));if(index<0)return false;
  const deaths=entries.filter(e=>e.status==='verified_death'&&e.victim===index).length,kills=entries.filter(e=>e.killer===index).length;
  const official=verifiedReplayScoreboard(match.players[index]);
  return a.deaths_match===true&&a.kills_match===true&&a.death_events===deaths&&a.kill_events===kills&&a.counter_deaths===deaths&&a.counter_kills===kills&&(!official||official.deaths===a.counter_deaths&&official.kills===a.counter_kills);
 });
 const complete=c.clock_complete===true&&c.game_end_observed===true&&c.truncated===false&&c.counter_resets===0&&c.raw_deaths===raw.entries.length&&invalid===0&&unresolved===0&&audit&&verifiedDeaths===c.verified_deaths&&verifiedKills===c.verified_kills&&reincarnations===c.verified_reincarnations;
 const cells=players.flatMap((a,killer)=>players.flatMap((b,victim)=>a.team!==b.team?[{killer,victim,count:entries.filter(e=>e.killer===killer&&e.victim===victim).length|| (complete?0:null),eventIds:entries.filter(e=>e.killer===killer&&e.victim===victim).map(e=>e.id)}]:[]));
 return {players,entries,cells,coverage:{complete,verifiedDeaths,verifiedKills,reincarnations,unresolved,nonHeroKills:entries.filter(e=>e.status==='verified_death'&&e.killer===null&&['allied_deny','self','environment'].includes(e.attribution)).length,invalid}};
}

File: tools/replay-parser/hero_death_journal.go
package main

import (
	"fmt"
	"github.com/dotabuff/manta"
	"github.com/dotabuff/manta/dota"
	"math"
	"sort"
	"strings"
)

const maxHeroJournalEvents = 4096
const maxHeroCounterSamples = 2000000

type HeroDeathCounter struct {
	From   float64 `json:"from"`
	To     float64 `json:"to"`
	Before int     `json:"before"`
	After  int     `json:"after"`
	Delta  int     `json:"delta"`
	Source string  `json:"source"`
}
type HeroDeathJournalEntry struct {
	VictimResourceSlot *int              `json:"victim_resource_slot"`
	KillerResourceSlot *int              `json:"killer_resource_slot"`
	ID                 string            `json:"id"`
	Seconds            float64           `json:"seconds"`
	VictimHero         *string           `json:"victim_hero"`
	VictimSteamID      *string           `json:"victim_steam_id"`
	VictimTeam         *string           `json:"victim_team"`
	RawTarget          string            `json:"raw_target"`
	RawAttacker        string            `json:"raw_attacker"`
	RawSource          *string           `json:"raw_source"`
	AttackerTeam       *string           `json:"attacker_team"`
	WillReincarnate    *bool             `json:"will_reincarnate"`
	Status             string            `json:"status"`
	DeathCounter       *HeroDeathCounter `json:"death_counter"`
	KillerHero         *string           `json:"killer_hero"`
	KillerSteamID      *string           `json:"killer_steam_id"`
	KillerTeam         *string           `json:"killer_team"`
	KillCounter        *HeroDeathCounter `json:"kill_counter"`
	Attribution        string            `json:"attribution"`
	SourceControlled   bool              `json:"source_controlled"`
	SourceIllusion     bool              `json:"source_illusion"`
}
type HeroDeathAudit struct {
	Hero          string `json:"hero"`
	DeathEvents   int    `json:"death_events"`
	CounterDeaths *int   `json:"counter_deaths"`
	KillEvents    int    `json:"kill_events"`
	CounterKills  *int   `json:"counter_kills"`
	DeathsMatch   *bool  `json:"deaths_match"`
	KillsMatch    *bool  `json:"kills_match"`
}
type HeroDeathJournal struct {
	Version  string                  `json:"version"`
	Entries  []HeroDeathJournalEntry `json:"entries"`
	Coverage struct {
		ClockSource            string           `json:"clock_source"`
		ClockComplete          bool             `json:"clock_complete"`
		GameEndObserved        bool             `json:"game_end_observed"`
		RawDeaths              int              `json:"raw_deaths"`
		EntriesStored          int              `json:"entries_stored"`
		VerifiedDeaths         int              `json:"verified_deaths"`
		VerifiedReincarnations int              `json:"verified_reincarnations"`
		VerifiedKills          int              `json:"verified_kills"`
		Unverified             int              `json:"unverified"`
		Contradictions         int              `json:"contradictions"`
		CounterResets          int              `json:"counter_resets"`
		Truncated              bool             `json:"truncated"`
		DroppedEvents          int              `json:"dropped_events"`
		DroppedCounterSamples  int              `json:"dropped_counter_samples"`
		FinalAudit             []HeroDeathAudit `json:"final_audit"`
	} `json:"coverage"`
}
type heroCounterSample struct {
	time          float64
	kills, deaths int
}
type heroCounterHistory struct {
	slot       int
	hero, team string
	steam      uint64
	samples    []heroCounterSample
	invalid    bool
}
type heroJournalRaw struct {
	ts                                                 float32
	target, attacker, source, targetTeam, attackerTeam string
	reincarnate                                        *bool
	illusion                                           bool
}
type heroDeathJournalCollector struct {
	clock                                            *wardLifetimeCollector
	histories                                        map[int]*heroCounterHistory
	events                                           []heroJournalRaw
	ended                                            bool
	endRaw                                           float64
	raw, dropped, sampleCount, sampleDropped, resets int
	clockMissing                                     bool
}

func newHeroDeathJournal(p *manta.Parser, clock *wardLifetimeCollector) *heroDeathJournalCollector {
	c := &heroDeathJournalCollector{clock: clock, histories: map[int]*heroCounterHistory{}}
	var resource *manta.Entity
	var rows []scoreboardCandidate
	dirty := false
	p.OnEntity(func(e *manta.Entity, op manta.EntityOp) error {
		if e.GetClassName() == "CDOTA_PlayerResource" {
			resource = e
			dirty = true
		}
		return nil
	})
	p.Callbacks.OnCSVCMsg_PacketEntities(func(_ *dota.CSVCMsg_PacketEntities) error {
		if c.ended || resource == nil {
			return nil
		}
		if dirty {
			rows = readReplayScoreboard(resource.Get, func(h uint64) string {
				if e := p.FindEntityByHandle(h); e != nil && strings.HasPrefix(e.GetClassName(), "CDOTA_Unit_Hero_") {
					return classToHero(e.GetClassName())
				}
				return ""
			})
			dirty = false
		}
		t := clock.clock(p.NetTick)
		if !lifetimeFinite(t) {
			if len(c.events) > 0 {
				c.clockMissing = true
			}
			return nil
		}
		for _, r := range rows {
			c.sample(r, t)
		}
		return nil
	})
	p.Callbacks.OnCMsgDOTACombatLogEntry(func(e *dota.CMsgDOTACombatLogEntry) error {
		if c.ended {
			return nil
		}
		if e.GetType() == dota.DOTA_COMBATLOG_TYPES_DOTA_COMBATLOG_GAME_STATE && e.GetValue() == 6 {
			c.ended = true
			c.endRaw = float64(e.GetTimestamp())
			return nil
		}
		name := func(i uint32) string { s, _ := p.LookupStringByIndex("CombatLogNames", int32(i)); return s }
		c.observeDeath(e, name)
		return nil
	})
	return c
}
func (c *heroDeathJournalCollector) sample(r scoreboardCandidate, t float64) {
	if c.ended || !lifetimeFinite(t) {
		return
	}
	h := c.histories[r.value.ResourceSlot]
	if h == nil {
		h = &heroCounterHistory{slot: r.value.ResourceSlot, hero: r.hero, team: r.team, steam: r.steam}
		c.histories[r.value.ResourceSlot] = h
	}
	if h.hero != r.hero || h.team != r.team || h.steam != r.steam {
		h.invalid = true
		return
	}
	if len(h.samples) > 0 {
		last := h.samples[len(h.samples)-1]
		if t < last.time || r.value.Kills < last.kills || r.value.Deaths < last.deaths {
			c.resets++
			h.invalid = true
			return
		}
		// Pauses freeze native time. A delayed replicated counter update at the same
		// native instant is a zero-width observed bracket, not a clock reset.
		if t == last.time && last.kills == r.value.Kills && last.deaths == r.value.Deaths {
			return
		}
	}
	if c.sampleCount >= maxHeroCounterSamples {
		c.sampleDropped++
		return
	}
	h.samples = append(h.samples, heroCounterSample{t, r.value.Kills, r.value.Deaths})
	c.sampleCount++
}
func (c *heroDeathJournalCollector) observeDeath(e *dota.CMsgDOTACombatLogEntry, name func(uint32) string) {
	if c.ended || e.GetType() != dota.DOTA_COMBATLOG_TYPES_DOTA_COMBATLOG_DEATH || !e.GetIsTargetHero() || e.GetIsTargetIllusion() || e.Timestamp == nil || !lifetimeFinite(float64(e.GetTimestamp())) {
		return
	}
	target := name(e.GetTargetName())
	if !strings.HasPrefix(target, "npc_dota_hero_") {
		return
	}
	c.raw++
	if len(c.events) >= maxHeroJournalEvents {
		c.dropped++
		return
	}
	r := heroJournalRaw{ts: e.GetTimestamp(), target: target, attacker: name(e.GetAttackerName()), reincarnate: e.WillReincarnate, illusion: e.GetIsAttackerIllusion()}
	if e.DamageSourceName != nil {
		r.source = name(e.GetDamageSourceName())
	}
	if e.TargetTeam != nil {
		r.targetTeam = wardTeam(e.GetTargetTeam())
	}
	if e.AttackerTeam != nil {
		r.attackerTeam = wardTeam(e.GetAttackerTeam())
	}
	c.events = append(c.events, r)
}
func heroJournalInside(ts float32, from, to float64) bool {
	return ts >= math.Nextafter32(float32(from), float32(math.Inf(-1))) && ts <= math.Nextafter32(float32(to), float32(math.Inf(1)))
}
func heroJournalBracket(a, b heroCounterSample, kill bool, start float64) *HeroDeathCounter {
	before, after := a.deaths, b.deaths
	if kill {
		before, after = a.kills, b.kills
	}
	return &HeroDeathCounter{a.time - start, b.time - start, before, after, after - before, "CDOTA_PlayerResource"}
}
func heroJournalIdentity(players []*Player, h *heroCounterHistory) *Player {
	p := uniqueWardHero(players, "npc_dota_hero_"+h.hero, h.team)
	if p == nil || p.SteamID != h.steam {
		return nil
	}
	return p
}
func (c *heroDeathJournalCollector) finish(players []*Player, start float64) *HeroDeathJournal {
	out := &HeroDeathJournal{Version: "hero-death-journal-v1", Entries: []HeroDeathJournalEntry{}}
	cov := &out.Coverage
	cov.ClockSource = "gamerules"
	cov.ClockComplete = c.clock != nil && !c.clock.clockGap && !c.clock.clockInvalid && c.clock.haveSample && !c.clockMissing && c.clock.clockSamplesDropped == 0
	cov.GameEndObserved = c.ended
	cov.RawDeaths = c.raw
	cov.DroppedEvents = c.dropped
	cov.DroppedCounterSamples = c.sampleDropped
	cov.CounterResets = c.resets
	cov.Truncated = c.dropped > 0 || c.sampleDropped > 0
	cov.FinalAudit = []HeroDeathAudit{}
	if !lifetimeFinite(start) || start < 0 {
		cov.DroppedEvents += len(c.events)
		cov.Truncated = true
		return out
	}
	events := []heroJournalRaw{}
	for _, r := range c.events {
		if c.ended && float64(r.ts) > c.endRaw {
			continue
		}
		events = append(events, r)
	}
	sort.SliceStable(events, func(i, j int) bool { return events[i].ts < events[j].ts })
	for i, r := range events {
		e := HeroDeathJournalEntry{ID: fmt.Sprintf("death-%d", i+1), Seconds: float64(r.ts) - start, RawTarget: r.target, RawAttacker: r.attacker, AttackerTeam: lifetimeSide(r.attackerTeam), WillReincarnate: r.reincarnate, Status: "unverified", Attribution: "unknown", SourceIllusion: r.illusion}
		if r.source != "" {
			e.RawSource = lifetimePtr(r.source)
		}
		if p := uniqueWardHero(players, r.target, r.targetTeam); p != nil {
			e.VictimHero = lifetimePtr(p.Hero)
			e.VictimSteamID = lifetimePtr(fmt.Sprint(p.SteamID))
			e.VictimTeam = lifetimeSide(p.Team)
		}
		out.Entries = append(out.Entries, e)
	}
	// Each positive counter delta must be exhausted by a unique set of event candidates.
	// No final total participates in this join.
	type match struct {
		indices []int
		bracket *HeroDeathCounter
		h       *heroCounterHistory
	}
	deathMatches := []match{}
	deathUses := make([]int, len(events))
	histories := []*heroCounterHistory{}
	for _, h := range c.histories {
		if !h.invalid && heroJournalIdentity(players, h) != nil {
			histories = append(histories, h)
		}
	}
	if cov.ClockComplete && !cov.Truncated {
		for _, h := range histories {
			for j := 1; j < len(h.samples); j++ {
				a, b := h.samples[j-1], h.samples[j]
				if b.deaths <= a.deaths {
					continue
				}
				indices := []int{}
				for i, r := range events {
					if normalizedHero(r.target) == normalizedHero(h.hero) && (r.targetTeam == h.team || r.targetTeam == "") && heroJournalInside(r.ts, a.time, b.time) {
						indices = append(indices, i)
					}
				}
				if len(indices) != b.deaths-a.deaths {
					continue
				}
				identitiesKnown := true
				for _, i := range indices {
					if out.Entries[i].VictimHero == nil || events[i].targetTeam != h.team {
						identitiesKnown = false
					}
				}
				if !identitiesKnown {
					continue
				}
				m := match{indices, heroJournalBracket(a, b, false, start), h}
				deathMatches = append(deathMatches, m)
				for _, i := range indices {
					deathUses[i]++
				}
			}
		}
		for _, m := range deathMatches {
			valid := true
			for _, i := range m.indices {
				if deathUses[i] != 1 {
					valid = false
				}
			}
			if !valid {
				continue
			}
			for _, i := range m.indices {
				e := &out.Entries[i]
				e.DeathCounter = m.bracket
				e.VictimResourceSlot = lifetimePtr(m.h.slot)
				if e.WillReincarnate != nil && *e.WillReincarnate {
					e.Status = "contradiction"
				} else {
					e.Status = "verified_death"
				}
			}
		}
		// An explicit return flag needs an independently observed unchanged counter bracket.
		for i, r := range events {
			e := &out.Entries[i]
			if e.Status != "unverified" || r.reincarnate == nil || !*r.reincarnate || e.VictimHero == nil {
				continue
			}
			for _, h := range histories {
				if normalizedHero(h.hero) != normalizedHero(*e.VictimHero) || h.team != r.targetTeam {
					continue
				}
				j := sort.Search(len(h.samples), func(j int) bool { return float32(h.samples[j].time) >= r.ts })
				if j > 0 && j < len(h.samples) {
					a, b := h.samples[j-1], h.samples[j]
					if a.deaths == b.deaths && heroJournalInside(r.ts, a.time, b.time) {
						e.Status = "verified_reincarnation"
						e.VictimResourceSlot = lifetimePtr(h.slot)
						e.DeathCounter = heroJournalBracket(a, b, false, start)
					}
				}
			}
		}
		killMatches := []match{}
		killUses := make([]int, len(events))
		for _, h := range histories {
			for j := 1; j < len(h.samples); j++ {
				a, b := h.samples[j-1], h.samples[j]
				if b.kills <= a.kills {
					continue
				}
				indices := []int{}
				for i, r := range events {
					e := &out.Entries[i]
					if strings.HasPrefix(r.source, "npc_dota_hero_") && normalizedHero(r.source) == normalizedHero(h.hero) && heroJournalInside(r.ts, a.time, b.time) {
						// Unknown victim/counter evidence cannot be silently removed to make
						// the remaining candidates fit a kill delta.
						if e.Status == "verified_reincarnation" {
							continue
						}
						if e.VictimTeam != nil && *e.VictimTeam == h.team && r.attackerTeam == h.team {
							continue
						}
						indices = append(indices, i)
					}
				}
				if len(indices) != b.kills-a.kills {
					continue
				}
				allVerified := true
				for _, i := range indices {
					e := out.Entries[i]
					if e.Status != "verified_death" || events[i].attackerTeam != h.team || e.VictimTeam == nil || *e.VictimTeam == h.team {
						allVerified = false
					}
				}
				if !allVerified {
					continue
				}
				killMatches = append(killMatches, match{indices, heroJournalBracket(a, b, true, start), h})
				for _, i := range indices {
					killUses[i]++
				}
			}
		}
		for _, m := range killMatches {
			valid := true
			for _, i := range m.indices {
				if killUses[i] != 1 {
					valid = false
				}
			}
			if !valid {
				continue
			}
			for _, i := range m.indices {
				e := &out.Entries[i]
				e.KillerResourceSlot = lifetimePtr(m.h.slot)
				e.KillerHero = lifetimePtr(heroJournalIdentity(players, m.h).Hero)
				e.KillerSteamID = lifetimePtr(fmt.Sprint(m.h.steam))
				e.KillerTeam = lifetimeSide(m.h.team)
				e.KillCounter = m.bracket
				e.Attribution = "verified_enemy_kill"
				rawAttacker := events[i].attacker
				e.SourceControlled = e.SourceIllusion || (rawAttacker != "" && rawAttacker != "dota_unknown" && rawAttacker != "unknown" && normalizedHero(rawAttacker) != normalizedHero(m.h.hero))
			}
		}
	}
	for i := range out.Entries {
		e := &out.Entries[i]
		r := events[i]
		if e.Status == "verified_death" && e.Attribution == "unknown" && e.VictimTeam != nil {
			if p := uniqueWardHero(players, r.source, r.attackerTeam); p != nil && p.Team == *e.VictimTeam {
				if e.VictimHero != nil && p.Hero == *e.VictimHero {
					e.Attribution = "self"
				} else {
					e.Attribution = "allied_deny"
				}
			} else if !strings.HasPrefix(r.attacker, "npc_dota_hero_") && r.source != "" && r.source == r.attacker && (strings.HasPrefix(r.attacker, "npc_dota_neutral_") || strings.HasPrefix(r.attacker, "npc_dota_goodguys_") || strings.HasPrefix(r.attacker, "npc_dota_badguys_") || r.attacker == "dota_fountain" || r.attacker == "npc_dota_roshan") {
				e.Attribution = "environment"
			}
		}
		switch e.Status {
		case "verified_death":
			cov.VerifiedDeaths++
		case "verified_reincarnation":
			cov.VerifiedReincarnations++
		case "contradiction":
			cov.Contradictions++
		default:
			cov.Unverified++
		}
		if e.Attribution == "verified_enemy_kill" {
			cov.VerifiedKills++
		}
	}
	cov.EntriesStored = len(out.Entries)
	for _, p := range players {
		a := HeroDeathAudit{Hero: p.Hero}
		for _, e := range out.Entries {
			if e.Status == "verified_death" && e.VictimHero != nil && *e.VictimHero == p.Hero {
				a.DeathEvents++
			}
			if e.Attribution == "verified_enemy_kill" && e.KillerHero != nil && *e.KillerHero == p.Hero {
				a.KillEvents++
			}
		}
		if c.ended && p.ReplayScoreboard != nil && p.ReplayScoreboard.Complete {
			a.CounterDeaths = lifetimePtr(p.ReplayScoreboard.Deaths)
			a.CounterKills = lifetimePtr(p.ReplayScoreboard.Kills)
			a.DeathsMatch = lifetimePtr(a.DeathEvents == *a.CounterDeaths)
			a.KillsMatch = lifetimePtr(a.KillEvents == *a.CounterKills)
		}
		cov.FinalAudit = append(cov.FinalAudit, a)
	}
	return out
}
